### DISCUSSION TITLE
Talk about the underlying buffer data structure of the BloomFilter

### DISCUSSION THREAD

**Discussion by zncleon** [apache/kvrocks#1800] [General]
Currently, the underlying buffer data structure of the BloomFilter is `nonstd::span<char>`[1], but in some scenes, it maybe not suitable. 
For example, the PinnableSlice `.data()` return the `const char*`, which can't fit the BlockSplitBloomFilter for the data type is `span<char>` and we need `span<const char>`. I have 2 solutions to temporarily resolve it:
1. Use template `span<T>` and specialize `span<char>` and `span<const char>`
2. Just use `const_cast<char *>` to converse the PinnableSlice `.data()` return 

But I think the two way is not graceful and reasonable. 
If we should introduce a widely suitable data structure[2], which ont only contains the pointer to the data but also contains the `is_mutable`, `is_owned` and other member to fit in more scenes.

---
[1] https://github.com/apache/kvrocks/blob/unstable/src/types/bloom_filter.h
[2] https://github.com/apache/arrow/blob/main/cpp/src/arrow/buffer.h

**Comment by mapleFU** [comment]
How about just hide the mutable inside the BloomFilter(without adding a `Buffer` type), and use `const_cast` internal? 🤔

Also cc @PragmaTwice
